### PR TITLE
Develop remove academy done

### PR DESCRIPTION
**Pull request by vincentpierre**
### Design
[Design Document](https://docs.google.com/document/d/179FpNmkAfl6-bud02ayyQ6ZzgeH921AxWRmmHC7AjCo/edit#)
### List of changes : 
 - Removed the Done flag of the Academy
 - Removed the MaxStep of the Academy
 - Removed/Modified the unit tests that tested for this functionality
 - Removed the `Academy.Done()` and `Academy.IsDone()` methods
 - Removed the `global_done` flag of the `BaseUnityEnvironment` abstract class
 - Added a ForceReset method on the Agent so the Academy can force all agents to reset when Python requests it
 - It is no longer necessary to call reset on the environment right after creating it (step will call reset if it is called for the first time)

### CODE REVIEW

File: protobuf-definitions/proto/mlagents/envs/communicator_objects/unity_rl_output.proto [hunk 1]
```diff
@@ -9,7 +9,6 @@ message UnityRLOutput {
     message ListAgentInfoProto {
         repeated AgentInfoProto value = 1;
     }
-    bool global_done = 1;
```
Comment (chriselion) [inline protobuf-definitions/proto/mlagents/envs/communicator_objects/unity_rl_output.proto]: Need to regenerate the associated code for this, right? This sounds like a better way to "delete" the field: https://developers.google.com/protocol-buffers/docs/proto#reserved

Comment (vincentpierre) [reply protobuf-definitions/proto/mlagents/envs/communicator_objects/unity_rl_output.proto]: Yes, but I think I should not regenerate the files at this stage so we have less diff files to look at. Also, it will make it easier to have parallel PRs on the proto definitions without messing up the generated files.

Comment (chriselion) [reply protobuf-definitions/proto/mlagents/envs/communicator_objects/unity_rl_output.proto]: Sounds good on not regenerating in this PR, but I would still suggest replacing this line with
```
reserved 1; // deprecated bool global_done field
```

**Comment by harperj** [comment]
Code looks good to me aside from fixing the minor linting issue from CircleCI.  You should add a reference to this change in the changelog.  It would also be nice to see a test that the "reset on first step" works correctly.

**Comment by vincentpierre** [comment]
DO NOT MERGE
Once the PR is approved, I will make  the commits for documentation changes.

**Comment by vincentpierre** [comment]
@chriselion I am rerequesting review since I edited the documentation.

**Comment by vincentpierre** [comment]
@chriselion I just did, thanks for catching this.

File: docs/Migrating.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
 #### Steps to Migrate
 * `UnitySDK/Assets/ML-Agents/Scripts/Communicator.cs` and its class `Communicator` have been renamed to `UnitySDK/Assets/ML-Agents/Scripts/ICommunicator.cs` and `ICommunicator` respectively.
 * The `SpaceType` Enums `discrete`, and `continuous` have been renamed to `Discrete` and `Continuous`.
+* We have removed the `Done` call on the Academy as well as the capacity to set `Max Steps`. Therefore, it is now impossible to call an AcademyReset from C# (only possible from Python). If you want to reset the simulation after a fixed number of steps, or when an event in the simulation occurs, we recommend looking at our multi-agent example environments (such as BananaCollector). In our examples, groups of Agents can be reset through an "Area" that can reset groups of Agents.
```
Comment (awjuliani) [inline docs/Migrating.md]: "as well as the capacity to set `Max Steps`" Makes it a little ambiguous as to whether max steps have also been removed from agents. I believe it would be better to rephrase this sentence a little bit.

File: UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs [hunk 3]
```diff
@@ -111,12 +110,12 @@ public void TestAgent()
 
     public class EditModeTestInitialization
     {
-/*
-        private Brain GenerateTestBrain()
-        {
-            return ScriptableObject.CreateInstance<TestBrain>();
-        }
-*/
+        /*
```
Comment (awjuliani) [inline UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs]: Why is this commented out, but not removed?

File: UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs [hunk 4]
```diff
@@ -344,6 +335,11 @@ public void TestAcademy()
                 }
 
                 stepsSinceReset += 1;
+<<<<<<< HEAD
```
Comment (awjuliani) [inline UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs]: Looks like something went wrong with a merge?

File: docs/Learning-Environment-Design-Academy.md [hunk 5]
```diff
@@ -31,7 +31,7 @@ in your Academy subclass.
 Implement an `AcademyReset()` function to alter the environment at the start of
 each episode. For example, you might want to reset an Agent to its starting
 position or move a goal to a random position. An environment resets when the
-Academy `Max Steps` count is reached.
+Python training process needs to reset the environment to facilitate training.
```
Comment (awjuliani) [inline docs/Learning-Environment-Design-Academy.md]: Maybe be more explicit here and say when `env.reset()` is called, since technically a reset could be called outside of the context of training.

File: UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs (hunk 4 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs]: Good catch!

File: UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs (hunk 3 above)
Comment (vincentpierre) [reply UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs]: Actually, only the tabs were changed. I do not know why this code is commented so I will remove it

File: docs/Learning-Environment-Design-Academy.md (hunk 5 above)
Comment (vincentpierre) [reply docs/Learning-Environment-Design-Academy.md]: Edited it.

File: UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs (hunk 4 above)
Comment (chriselion) [reply UnitySDK/Assets/ML-Agents/Editor/Tests/MLAgentsEditModeTest.cs]: There's a pre-commit check we can add for this. Remind me later.

File: docs/Migrating.md [hunk 6]
```diff
@@ -8,6 +8,7 @@
 #### Steps to Migrate
 * `UnitySDK/Assets/ML-Agents/Scripts/Communicator.cs` and its class `Communicator` have been renamed to `UnitySDK/Assets/ML-Agents/Scripts/ICommunicator.cs` and `ICommunicator` respectively.
 * The `SpaceType` Enums `discrete`, and `continuous` have been renamed to `Discrete` and `Continuous`.
+* We have removed the `Done` call as well as the capacity to set `Max Steps` on the Academy. Therefore, it is now impossible to call an AcademyReset from C# (only possible from Python). If you want to reset the simulation after a fixed number of steps, or when an event in the simulation occurs, we recommend looking at our multi-agent example environments (such as BananaCollector). In our examples, groups of Agents can be reset through an "Area" that can reset groups of Agents.
```
Comment (chriselion) [inline docs/Migrating.md]: > Therefore, it is now impossible to call an AcademyReset from C# (only possible from Python). 

It's still possible for the user to call it from "game" code, right? Maybe rephrase as something like "Therefore an AcademyReset will never be triggered from C# (only from Python)"?
